server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java

```diff
@@ -112,8 +116,11 @@ protected String contentType() {
     public void postParse(DocumentParserContext context) {
         // Ignored values are only expected in synthetic mode.
         assert context.getIgnoredFieldValues().isEmpty() || context.mappingLookup().isSourceSynthetic();
-        for (NameValue nameValue : context.getIgnoredFieldValues()) {
-            context.doc().add(new StoredField(NAME, encode(nameValue)));
+        List<NameValue> ignoredFieldValues = new ArrayList<>(context.getIgnoredFieldValues());
+        // ensure consistent ordering when retrieving synthetic source
+        Collections.sort(ignoredFieldValues, Comparator.comparing(NameValue::name));
+        for (NameValue nameValue : ignoredFieldValues) {
+            nameValue.doc().add(new StoredField(NAME, encode(nameValue)));
```

Comment:
Sorry I missed where doc() doesn't map to context.doc(), any quick pointer here?

Reply:
See https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java#L556 where we switch the current doc to a new nested doc.
For each nested doc we have a corresponding Lucene document and the root object is the first one.  So we need to add the ignored fields in the right Lucene document depending on where they are registered (in the root or within a nested document).

Reply:
I see, it's pre-existing.